src/app/3d/qgs3dmaptoolmeasureline.cpp

```diff
@@ -0,0 +1,304 @@
+  {
+    if ( mMeasureLineTool->done() )
+    {
+      mMeasureLineTool->mDialog->restart();
+    }
+
+    // Left button, keep addinng point
+    qInfo() << "handlePickOnVectorLayer" ;
+    QgsVector3D mapCoords = Qgs3DUtils::worldToMapCoordinates(
+                              QgsVector3D( worldIntersection.x(),
+                                           worldIntersection.y(),
+                                           worldIntersection.z() ), mMeasureLineTool->mCanvas->map()->origin() );
+    QgsPoint pt( mapCoords.x(), mapCoords.y(), mapCoords.z() );
+    qInfo() << "Coord (handlePickOnVectorLayer): " << pt.x() << " " << pt.y() << " " << pt.z();
+
+    mMeasureLineTool->mDone = false;
+    mMeasureLineTool->addPoint( pt );
+    mMeasureLineTool->mDialog->show();
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::RightButton )
+  {
+    // Finish measurement
+    qInfo() << "Finish measurement";
+    mMeasureLineTool->mDone = true;
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::MiddleButton )
+  {
+    qInfo() << "Undo - middle button";
+    mMeasureLineTool->undo();
+
+  }
+}
+
+Qgs3DMapToolMeasureLine::Qgs3DMapToolMeasureLine( Qgs3DMapCanvas *canvas )
+  : Qgs3DMapTool( canvas )
+{
+  qInfo() << "Constructed";
+  connect( mCanvas->scene(), &Qgs3DMapScene::terrainEntityChanged, this, &Qgs3DMapToolMeasureLine::onTerrainEntityChanged );
+  mPickHandler.reset( new Qgs3DMapToolMeasureLinePickHandler( this ) );
+
+  // Line style
+  mLineSymbol = new QgsLine3DSymbol;
+  mLineSymbol->setRenderAsSimpleLines( true );
+  mLineSymbol->setWidth( 4 );
+  QgsPhongMaterialSettings phongMaterial;
+  phongMaterial.setAmbient( Qt::yellow );
+  phongMaterial.setDiffuse( Qt::green );
+  mLineSymbol->setMaterial( phongMaterial );
+
+  mLineSymbolRenderer = new QgsVectorLayer3DRenderer( mLineSymbol );
+
+  // Dialog
+  mDialog = new Qgs3DMeasureDialog( this );
+  mDialog->setWindowFlags( mDialog->windowFlags() | Qt::Tool );
+  mDialog->restorePosition();
+}
+
+Qgs3DMapToolMeasureLine::~Qgs3DMapToolMeasureLine() = default;
+
+void Qgs3DMapToolMeasureLine::activate()
+{
+  qInfo() << "Measure line activated";
+  if ( QgsTerrainEntity *terrainEntity = mCanvas->scene()->terrainEntity() )
+  {
+    connect( terrainEntity->terrainPicker(), &Qt3DRender::QObjectPicker::clicked, this, &Qgs3DMapToolMeasureLine::onTerrainPicked );
+  }
+
+  mCanvas->scene()->registerPickHandler( mPickHandler.get() );
+
+  // Initialize the measurement line
+  mMeasurementLine = new QgsLineString();
+  mMeasurementLine->addZValue();
+
+  // Initialize measurement feature
+  mMeasurementFeature = new QgsFeature( QgsFeatureId( 1 ) );
+  mMeasurementFeature->setGeometry( QgsGeometry( mMeasurementLine ) );
+
+  // Initialize the line layer
+  QString mapCRS = mCanvas->map()->crs().authid();
+  mMeasurementLayer = new QgsVectorLayer( QStringLiteral( "LineStringZ?crs=" ) + mapCRS, QStringLiteral( "Measurement" ), QStringLiteral( "memory" ) );
```

Comment:
maybe better to create the measurement layer only when the tool is activated for the first time. Subsequent activations could just "reset" the feature geometry like in restart(). Otherwise we may be leaving some memory leaks of vector layers after each activation.

Reply:
done